src/task/task_hash.rs

```diff
@@ -0,0 +1,168 @@
+
+        for script in activation_scripts {
+            script.hash(&mut hasher);
+        }
+
+        let mut urls = Vec::new();
+
+        if let Some(env) = lock_file.environment(run_environment.name().as_str()) {
+            if let Some(packages) = env.packages(Platform::current()) {
+                for package in packages {
+                    urls.push(match package {
+                        Package::Conda(c) => c.url().clone(),
+                        Package::Pypi(p) => p.url().clone(),
+                    })
+                }
+            }
+        }
+
+        urls.sort();
+
+        urls.hash(&mut hasher);
+        EnvironmentHash(format!("{:x}", hasher.finish()))
+    }
+}
+
+/// The [`TaskHash`] group all the hashes of a task. It can be converted to a [`ComputationHash`]
+/// with the [`TaskHash::computation_hash`] method.
+#[derive(Debug)]
+pub struct TaskHash {
+    pub environment: EnvironmentHash,
+    pub command: Option<String>,
+    pub inputs: Option<InputHashes>,
+    pub outputs: Option<OutputHashes>,
+}
+
+impl TaskHash {
+    /// Constructs an instance from an executable task.
+    pub async fn from_task(
+        task: &ExecutableTask<'_>,
+        lock_file: &LockFile,
+    ) -> Result<Option<Self>, InputHashesError> {
+        let input_hashes = InputHashes::from_task(task).await?;
+        let output_hashes = OutputHashes::from_task(task).await?;
+
+        if input_hashes.is_none() && output_hashes.is_none() {
+            return Ok(None);
+        }
+
+        Ok(Some(Self {
+            command: task.full_command(),
+            outputs: output_hashes,
+            inputs: input_hashes,
+            environment: EnvironmentHash::from_environment(&task.run_environment, lock_file),
+        }))
+    }
+
+    pub async fn update_output(
+        &mut self,
+        task: &ExecutableTask<'_>,
+    ) -> Result<(), InputHashesError> {
+        self.outputs = OutputHashes::from_task(task).await?;
+        Ok(())
+    }
+
+    /// Computes a single hash for the task.
+    pub fn computation_hash(&self) -> ComputationHash {
+        let mut hasher = Xxh3::new();
+        self.command.hash(&mut hasher);
+        self.inputs.hash(&mut hasher);
+        self.outputs.hash(&mut hasher);
+        self.environment.hash(&mut hasher);
+        ComputationHash(format!("{:x}", hasher.finish()))
+    }
+}
+
+/// The combination of all the hashes of the inputs of a task.
+#[derive(Debug, Hash)]
+pub struct InputHashes {
+    pub files: FileHashes,
+}
+
+impl InputHashes {
+    /// Compute the input hashes from a task.
+    pub async fn from_task(task: &ExecutableTask<'_>) -> Result<Option<Self>, InputHashesError> {
+        let Some(ref inputs) = task.task().as_execute().and_then(|e| e.inputs.clone()) else {
+            return Ok(None);
+        };
+
+        let files =
+            FileHashes::from_files(&task.working_directory().unwrap(), inputs.iter()).await?;
```

Comment:
This will unwrap when a user accidentally add as file instead of a dir in the `cwd`